--- ./sox-effect-0.0.1/src/Makefile.am.gege	Mon Apr  9 17:33:43 2001
+++ ./sox-effect-0.0.1/src/Makefile.am	Mon Apr  9 17:36:41 2001
@@ -5,8 +5,7 @@
 libsox_la_LDFLAGS = -module -avoid-version
 LIBS = @GTK_LIBS@ @PTHREAD_LIBS@
 
-COMMON_CFLAGS = -O2 -I$(top_builddir) -I$(top_srcdir) -I$(top_srcdir)/.. -D_REENTRANT @GTK_CFLAGS@
+INCLUDES = -I$(top_builddir) -I$(top_srcdir) -I$(top_srcdir)/.. -D_REENTRANT @GTK_CFLAGS@
 COMMON_SRC = sox.c sox_configure.c avg.c band.c chorus.c copy.c cut.c deemphas.c dyn.c echo.c echos.c flanger.c highp.c lowp.c map.c mask.c phaser.c pick.c polyphas.c rate.c resample.c reverb.c reverse.c split.c stat.c swap.c vibro.c libst.c handlers.c misc.c util.c libst.h resampl.h sox.h version.h patchlvl.h resdefs.h st.h
 
-CFLAGS = $(COMMON_CFLAGS)
 libsox_la_SOURCES =  $(COMMON_SRC)
