--- ./xmms-shell-0.99.3/src/getline.cc.gege	2002-10-23 01:48:03.000000000 +0200
+++ ./xmms-shell-0.99.3/src/getline.cc	2003-07-08 17:30:55.000000000 +0200
@@ -1,7 +1,6 @@
 #include "config.h"
 #include "getline.h"
 #include <stdio.h>
-#include <string>
 
 #ifdef HAVE_LIBREADLINE
 
@@ -10,6 +9,7 @@
 #include <readline/readline.h>
 #include <readline/history.h>
 #include "command.h"
+#include <string>
 
 typedef char *(*CompletionHelper)(char **argv, int arg, int state);
 
